src/coq/decompiler/decompiler.ml

```diff
@@ -63,14 +92,25 @@ let run_tac env sigma (tac : unit Proofview.tactic) (goal : constr)
     
 (* Returns true if the given tactic solves the goal. *)
 let solves env sigma (tac : unit Proofview.tactic) (goal : constr) =
-  fst (run_tac env sigma tac goal) = []
+  let subgoals, _ = run_tac env sigma tac goal in subgoals = []
 
 (* Compute the type of a term if possible, otherwise None. *)
 let type_of env (trm : constr) : types option =
   try Some (Typeops.infer env trm).uj_type
   with _ -> None
-  
-            
+
+let subterm_with_type env sigma : types -> constr -> (Evd.evar_map * (env * constr)) list =
```

Comment:
I would name this `subterms_with_types`. And I think it should go in either `hofs.ml` or `typehofs.ml`, since it is useful more generally :).

Reply:
Moved to typehofs.ml